### PR TITLE
[#137] Add headers to Trailing* hints

### LINKED ISSUE #137: Add headers to Trailing* hints
If we submit HLint hints, we write additional header like:

```haskell
Warning: Redundant do
```

I propose to add this header to all our hints. So when trailing newline are removed, we can write something like:

```haskell
Trailing newline
```

(text is subject to discussion)

We probably can even have a data type like:

```haskell
data Suggestion = Suggestion
    { suggestionHeader :: !Text  -- ^ Hint text
    , suggestionBody :: !Text  -- ^ New line body
    }
```

and a function like:

```haskell
formatSuggestion :: Suggestion -> Text
formatSuggestion Suggestion{..} = unlines
    [ suggestionHead
    , "```suggestion"
    , suggestionBody
    , "```"
    ]
```

### PR DESCRIPTION
**Pull request by vrom911**
Resolves #137

There is a lot of space for improvements. For example, we could add `Severity` to that data type. or to specify `HLint` for HLint hints, maybe other interesting thoughts...

### CODE REVIEW

File: src/Hintman/Core/Hint.hs [hunk 1]
```diff
@@ -10,6 +15,27 @@ module Hintman.Core.Hint
 import Relude.Extra.Enum (inverseMap)
 
 
+-- | Data type to represent the information for comment creation.
+data Hint = Hint
+    { hintHeader       :: !Text  -- ^ Hint text
+    , hintBody         :: !Text  -- ^ New line body
+    , hintIsSuggestion :: !Bool  -- ^ Is this GitHub @suggestion@?
```
Comment (chshersh) [inline src/Hintman/Core/Hint.hs]: Am I right that we will get rid of this field when GitHub API supports multiline suggestions?

File: src/Hintman/Core/Hint.hs [hunk 2]
```diff
@@ -10,6 +15,27 @@ module Hintman.Core.Hint
...
+
+-- | Formats 'Hint' in the compatible with @GitHub@ way.
+formatHint :: Hint -> Text
+formatHint Hint{..} = unlines $ concat
+    [ [ hintHeader | hintHeader /= ""]
+    , ["```" <> if hintIsSuggestion then "suggestion" else ""]
+    , [ hintBody | hintType /= TrailingNewlines ]
+    , [ "```"]
+    , [ "__Note:__ " <> hintNote | hintNote /= "" ]
+    ]
```
Comment (chshersh) [inline src/Hintman/Core/Hint.hs]: Wow, this is actually very cool to write this function and take all cases into consideration :open_mouth: :boom:

File: src/Hintman/Core/Review.hs [hunk 3]
```diff
@@ -45,15 +45,12 @@ data Comment = Comment
...
-    , commentHintType :: !HintType
+    , commentHint     :: !Hint
     } deriving (Show, Eq)
 
 instance ToJSON Comment where
     toJSON Comment{..} = object
         [ "path"     .= commentPath
         , "position" .= commentPosition
-        , "body"     .= commentBody
+        , "body"     .= formatHint commentHint
```
Comment (chshersh) [inline src/Hintman/Core/Review.hs]: Nice! I had the same idea in my mind :+1:

File: test/Test/Hint.hs [hunk 4]
```diff
@@ -12,6 +14,81 @@ import Test.Hint.TrailingSpaces (trailingSpacesSpec)
...
+hintFormatSpec :: Spec
+hintFormatSpec = describe "Hint formatted correctly" $ do
+    it "formats suggestion correctly" $
+        formatHint h1 `shouldBe` h1Txt
+    it "formats non-suggestion correctly" $
+        formatHint h2 `shouldBe` h2Txt
+    it "adds note" $
+        formatHint h3 `shouldBe` h3Txt
+    it "handles TrailingNewlines correctly" $
+        formatHint h4 `shouldBe` h4Txt
```
Comment (chshersh) [inline test/Test/Hint.hs]: Wow, these tests are just :sweat_drops: :sweat_drops: :sweat_drops:

File: src/Hintman/Core/Hint.hs (hunk 1 above)
Comment (vrom911) [reply src/Hintman/Core/Hint.hs]: Yes, this is correct :wink:

**Pull request merged**